File: paper_artifacts/experiments/llr8/artifacts/sources/oss120b-fortran-skills/fuse_move_ifs__submitted.f90
subroutine fuse_move_ifs_fp64(a, b, cond, src, K, LEN_2D) bind(C)
  use iso_c_binding
  implicit none

  integer(c_int64_t), value, intent(in) :: K, LEN_2D
  real(c_double), intent(inout) :: a(LEN_2D, LEN_2D)
  real(c_double), intent(inout) :: b(LEN_2D, LEN_2D)
  real(c_double), intent(in) :: cond(LEN_2D)
  real(c_double), intent(in) :: src(LEN_2D, LEN_2D)

  integer(c_int64_t) :: i, j

  !$omp parallel do private(j) schedule(static)
  do i = 1, LEN_2D
    !$omp simd
    do j = 1, LEN_2D
      if (cond(i) > 0.0_c_double) then
        a(j, i) = src(j, i) * 2.0_c_double
      end if
      if (K > 0) then
        b(j, i) = src(j, i) + 1.0_c_double
      end if
    end do
  end do
  !$omp end parallel do

end subroutine fuse_move_ifs_fp64

File: paper_artifacts/experiments/llr9/artifacts/sources/oss120b-c/fuse_move_ifs__submitted.c
/* Optimized version of fuse_move_ifs kernel.
 * This implementation fuses the two loop nests when K > 0, reducing memory traffic.
 * It also uses OpenMP parallelism and SIMD directives for better vectorization.
 */

#include <stdint.h>
#include <stddef.h>

void fuse_move_ifs_fp64(double *restrict a, double *restrict b, const double *restrict cond, const double *restrict src,
                        const int64_t K, const int64_t LEN_2D) {
    // Compute whether we need to compute b at all.
    const int doK = (K > 0);

    if (doK) {
        // Fuse the two loops: for each row, conditionally compute a and always compute b.
        // Parallelize over rows. Inner loop is vectorizable.
        #pragma omp parallel for schedule(static)
        for (int64_t i = 0; i < LEN_2D; ++i) {
            const int cond_pos = (cond[i] > 0.0);
            int64_t row_offset = i * LEN_2D;
            // Use SIMD for the inner loop.
            #pragma omp simd
            for (int64_t j = 0; j < LEN_2D; ++j) {
                double src_val = src[row_offset + j];
                if (cond_pos) {
                    a[row_offset + j] = src_val * 2.0;
                }
                b[row_offset + j] = src_val + 1.0;
            }
        }
    } else {
        // K <= 0: only compute a where cond is positive.
        #pragma omp parallel for schedule(static)
        for (int64_t i = 0; i < LEN_2D; ++i) {
            if (cond[i] > 0.0) {
                int64_t row_offset = i * LEN_2D;
                #pragma omp simd
                for (int64_t j = 0; j < LEN_2D; ++j) {
                    a[row_offset + j] = src[row_offset + j] * 2.0;
                }
            }
        }
    }
}
